include/highfive/bits/H5Reference_misc.hpp

```diff
@@ -34,21 +34,38 @@ inline void Reference::create_ref(hobj_ref_t* refptr) const {
 
 inline ObjectType Reference::getType(const Object& location) const {
     hid_t res;
+#if (H5Rdereference_vers == 2)
+    if ((res = H5Rdereference(location.getId(), H5P_DEFAULT, H5R_OBJECT, &href)) < 0) {
+        HDF5ErrMapper::ToException<ReferenceException>("Unable to dereference.");
+    }
+#else
     if ((res = H5Rdereference(location.getId(), H5R_OBJECT, &href)) < 0) {
         HDF5ErrMapper::ToException<ReferenceException>("Unable to dereference.");
     }
+#endif
     return Object(res).getType();
 }
 
 template <typename T>
 inline T Reference::dereference(const Object& location) const {
-    static_assert(std::is_base_of<Object, T>::value != 0,
-                  "Can only return a subtype of HighFive::Object");
+    static_assert(std::is_same<DataSet, T>::value || std::is_same<Group, T>::value,
+                  "We can only (de)reference HighFive::Group or HighFive:DataSet");
     hid_t res;
+#if (H5Rdereference_vers == 2)
+    if ((res = H5Rdereference(location.getId(), H5P_DEFAULT, H5R_OBJECT, &href)) < 0) {
+        HDF5ErrMapper::ToException<ReferenceException>("Unable to dereference.");
+    }
+#else
     if ((res = H5Rdereference(location.getId(), H5R_OBJECT, &href)) < 0) {
         HDF5ErrMapper::ToException<ReferenceException>("Unable to dereference.");
     }
-    return T(Object(res));
+#endif
+    Object obj(res);
```

Comment:
would it be too hard to enable T(res)? I guess we would need to make Group and Dataset friendly with us, but it seems alright with me and we avoid this double conversion

Reply:
I still need to construct first an object to get the type, so it seems redundant to me? otherwise I would agree with you that I've been doing this a bit roundabout.. maybe in this case it's ok (taking into account my dedup changes :) )